odoo/addons/base/models/res_partner.py

```diff
@@ -425,8 +425,9 @@ def onchange_company_type(self):
 
     @api.constrains('barcode')
     def _check_barcode_unicity(self):
-        if self.barcode and self.env['res.partner'].search_count([('barcode', '=', self.barcode)]) > 1:
-            raise ValidationError('An other user already has this barcode')
+        for partner in self:
+            if partner.barcode and self.env['res.partner'].search_count([('barcode', '=', partner.barcode)]) > 1:
```

Comment:
@ryv-odoo
`search_count` in a loop is not ideal, but may not be the performance bottleneck in the business
But I am curious in https://github.com/odoo/odoo/pull/41658, if the unique constraint is also used for search before the PR.
maybe we can index the company dependent field in ir.property with the help of the our new index ?@xmo-odoo

Reply:
I wouldn't do premature optimisation here. In 99,9% of cases, `self` will be a list of only 1 record. This PR fixes an extreme edge case.

Reply:
> irproperty already indexes all relevant fields

I mean when you search a `partner.search(['barcode', '=', 'foo'])`
The sql would be something like
```
SELECT "ir_property"."id" FROM "ir_property" WHERE ((("ir_property"."fields_id" = 890) AND (("ir_property"."company_id" in (1)) OR "ir_property"."company_id" IS NULL)) AND ("ir_property"."value_text" = 'foo')) ORDER BY  "ir_property"."id"
```
`name` and `res_id` are indexed in the `ir.property` but not the `value_text`

I just check with the pos team, at least for know they don't receive any performance complain about the barcode search for business. So it is OK for now.

Reply:
Indeed @HydrionBurst , but this code has four years and nobody complains about it. I am not a huge fan to change more than necessary in stable, it is why I didn't ask to change that.